workflows/ingest/masters.go

```diff
@@ -160,18 +110,84 @@ func uploadMaster(ctx workflow.Context, params MasterParams) (*MasterResult, err
 
 	createPreviewsAsync(ctx, []string{result.AssetID})
 
-	err = notifyImportCompleted(asyncCtx, params.Targets, params.Metadata.JobProperty.JobID, map[string]paths.Path{
-		result.AssetID: file,
-	})
+	return result.AssetID, nil
+}
+
+func uploadMaster(ctx workflow.Context, params MasterParams) (*MasterResult, error) {
+	var filename string
+	var err error
+	switch params.OrderForm {
+	case OrderFormOtherMaster, OrderFormVBMaster, OrderFormSeriesMaster, OrderFormLEDMaterial, OrderFormPodcast:
+		filename, err = masterFilename(params.Metadata.JobProperty)
+	case OrderFormVBMasterBulk:
+		break
+	default:
+		return nil, fmt.Errorf("unsupported order form: %s", params.OrderForm)
+	}
 
 	if err != nil {
 		return nil, err
 	}
 
+	sourceFiles := []paths.Path{}
+
+	if params.SourceFile != nil {
+		sourceFiles = append(sourceFiles, *params.SourceFile)
+	} else {
+		files, err := wfutils.ListFiles(ctx, params.Directory)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(files) == 0 {
+			return nil, fmt.Errorf("no files in directory: %s", params.Directory)
+		}
+
+		if len(files) > 1 && params.OrderForm != OrderFormVBMasterBulk {
+			return nil, fmt.Errorf("too many files in directory: %s", params.Directory)
+		}
+
+		sourceFiles = files
+	}
+
+	importedVXs := map[string]paths.Path{}
+	errors := []error{}
+
+	for _, sourceFile := range sourceFiles {
+		file := params.OutputDir.Append(sourceFile.Base())
+
+		if filename != "" {
+			file = params.OutputDir.Append(filename)
```

Comment:
How I'm reading this, we set filename only if `OrderFormOtherMaster, OrderFormVBMaster, OrderFormSeriesMaster, OrderFormLEDMaterial, OrderFormPodcast`.
And this condition right here looks like it is assuming sourceFiles will only have 1 file in that scenario, and with that assumption in mind we can safely do this (`file = params.OutputDir.Append(filename)`).
But it's hard to follow along on this and could have bugs if sourcefiles somehow got multiple items.

You could split the "bulk" scenario into a different function as part of the switch above, and split the shared code into functions instead.